internal-packages/workflow-designer-ui/src/editor/hooks/use-keyboard-shortcuts.ts

```diff
@@ -0,0 +1,114 @@
+import { useWorkflowDesigner } from "@giselle-sdk/giselle/react";
+import { useKeyPress } from "@xyflow/react";
+import { useCallback, useEffect, useRef } from "react";
+import { useNodeManipulation } from "../node";
+import {
+	moveTool,
+	selectActionTool,
+	selectLanguageModelTool,
+	selectRetrievalCategoryTool,
+	selectSourceCategoryTool,
+	selectTriggerTool,
+	useToolbar,
+} from "../tool/toolbar";
+
+// Browser shortcuts that should be prevented when canvas is focused
+const BROWSER_SHORTCUTS_TO_PREVENT = [
+	{ key: "d", modifiers: ["meta", "ctrl"] }, // Bookmark
+	// Add more shortcuts here as needed:
+	// { key: 's', modifiers: ['meta', 'ctrl'] }, // Save page
+	// { key: 'p', modifiers: ['meta', 'ctrl'] }, // Print
+];
+
+// Custom hook for handling key actions with repeat prevention
+function useKeyAction(
+	key: string | string[],
+	action: () => void,
+	enabled: boolean = false,
+) {
+	const wasPressed = useRef(false);
+	const isPressed = useKeyPress(key, { actInsideInputWithModifier: false });
+
+	useEffect(() => {
+		if (isPressed && !wasPressed.current && enabled) {
+			action();
+		}
+		wasPressed.current = isPressed;
+	}, [isPressed, enabled, action]);
+}
+
+export function useKeyboardShortcuts() {
+	const toolbar = useToolbar();
+	const { data } = useWorkflowDesigner();
+	const {
+		copy: handleCopy,
+		paste: handlePaste,
+		duplicate: handleDuplicate,
+	} = useNodeManipulation();
+
+	// Only use keyboard shortcuts when canvas is focused
+	const canUseShortcuts = data.ui.focusedArea === "canvas";
```

Comment:
I think that using isCanvasFocused will allow the code to express its intent without comments.

Reply:
Fixed [b7e3dbc](https://github.com/giselles-ai/giselle/pull/1590/commits/b7e3dbca41c565070cead9a6376a478b0029a2a8)